core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -559,6 +560,30 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  case class LeaderNode(leaderId: Int, leaderEpoch: Int, node: Node)
+
+  private def getCurrentLeader(tp: TopicPartition): LeaderNode = {
+    val partitionInfoOrError = replicaManager.getPartitionOrError(tp)
+    var leaderId = -1
+    var leaderEpoch = -1
+    partitionInfoOrError match {
+      case Right(x) =>
+          leaderId = x.leaderReplicaIdOpt.getOrElse(-1)
+          leaderEpoch = x.getLeaderEpoch
+      case Left(x) =>
+        debug(s"Unable to retrieve local leaderId and Epoch with error $x, falling back to metadata cache")
+        val partitionInfo = metadataCache.getPartitionInfo(tp.topic, tp.partition)
```

Comment:
any chance partitionInfo can be null?

Reply:
I don't think so, getPartitionInfo returns an Option, the equivalent of null would be an empty option.

Reply:
I see